### ISSUE TITLE
Implement gapless playback for looped ringtones

### ISSUE THREAD

**Issue by ltguillaume** [BlackyHawky/Clock#129]
**Is your feature request related to a problem? Please describe.**

Ever since Android 12, it seems that many clock apps have an issue with looped playback: files that previously looped gaplessly, now have a gap introduced between every loop.

**Describe the solution you'd like**

Since many ringtones sound strange with a gap between loops, it would be great if gapless playback can be reintroduced.

There's only one clock app I've found so far that *does* perfectly play gapless loops on Android 12+: https://github.com/gabeg805/NFC-Alarm-Clock/

It seems that it uses ExoPlayer for playback, perhaps this is the cause for that? (https://github.com/gabeg805/NFC-Alarm-Clock/blob/main/app/src/main/kotlin/com/nfcalarmclock/system/mediaplayer/NacMediaPlayer.kt)

**Comment by BlackyHawky** [comment]
> Since many ringtones sound strange...

Strange because I don't notice this on my devices with Android 12 and 14. 

> ... with a gap between loops

There's only a single second between 2 ringtone launches and I don't find it that bothersome. 🤔

> it would be great if gapless playback can be reintroduced.

Are you sure this has already been introduced? If so, I don't know about it.

In any case, I'm not sure I want to add new dependencies just for a short second between 2 ringtones...

**Comment by ltguillaume** [comment]
Nah, I understand that it's going too far to add new dependencies.

I just don't understand what happened after Android 12+. Nice little ringtones like the following one just stopped... well, being nice 😉

[Anticipation.zip](https://github.com/user-attachments/files/17807019/Anticipation.zip)

**Comment by BlackyHawky** [comment]
It's true that with the ringtone you've attached it's not very pretty. 😅
Perhaps using ExoPlayer would also solve #98.

To be tested in the hope that it will be easy to implement...

**Comment by BlackyHawky** [comment]
@ltguillaume 
Can you tell me if it's the same with [version 2.21](https://github.com/BlackyHawky/Clock/releases/tag/2.21)?

**Comment by ltguillaume** [comment]
Yeah, still the same 😭

**Comment by BlackyHawky** [comment]
@ltguillaume 
Hi,

Can you please test the apk below?
I've replaced `MediaPlayer` with `ExoPlayer` to play ringtones.

Note: the size of the application has increased by 2 MB... ☹️

[Clock_2.22-beta-debug.zip](https://github.com/user-attachments/files/20453648/Clock_2.22-beta-debug.zip)

**Comment by ltguillaume** [comment]
Yes, this does work perfectly now!

It's up to you to decide if the increase in size is worth it. Maybe there are other upsides (or downsides) for using ExoPlayer, like wider format support, more/less memory use, changes in stability (these are just possible differences, I haven't checked them).

**Comment by BlackyHawky** [comment]
@ltguillaume 
Personally, I don't see any harm in using `ExoPlayer` instead of `MediaPlayer`.

The only downside is the increased size of the app; I didn't expect that much. 😅
Anyway, less than 15MB seems reasonable (with this change the size of the release version will be 10.61 MB).

On the other hand, there some advantages that I do not consider negligible:
- Code is simplified;
- No additional permission is required;
- Automatic routing to Bluetooth devices if connected; otherwise to the speaker;
- Volume control whether a Bluetooth device is connected or not;
- And obviously, it doesn't introduce a pause or silence between repeat cycles;

That's why I think I'll validate the changes made today on my computer because it resolves this topic as well as #289. 😉

**Comment by ltguillaume** [comment]
Ah great, sounds to me like there are enough reasons to adopt this change then. Thanks for all the work! I'll close this issue then.

**Comment by BlackyHawky** [comment]
The commit I'll publish will close this topic automatically. 😉
